### ISSUE TITLE
`'Valid` for enumerations

### ISSUE THREAD

**Issue by mosteo** [Ada-Rapporteur-Group/User-Community-Input#95]
Would a `'Valid` for an enumeration type that took a [[w_]w_]string would be interesting? To be able to know in advance  if a `'Value` is going to succeed or raise.

You can roll out your own with generics, of course, but I find this more often that I would have thought when dealing with user input.

**Comment by Richard-Wai** [comment]
I have definitely found myself wanting of this! It's particularly useful when you're trying to map enumeration values to hardware registers, or sharing enumerations across languages.

**Comment by swbaird** [comment]
I understand that this was just a preliminary sketch of a proposal, but that won't stop me from (probably prematurely) commenting on  details.
I think that extending the definition of the existing "Valid" attribute would be mistake.
I think "Valid_Image", or "Is_Image", or even "Fred" would be a better name for this attribute.

**Comment by joshua-c-fletcher** [comment]
between overloading 'Valid or using 'Valid_Image, 'Is_Image, or 'Fred,
I prefer 'Valid_Image for the name of the proposed attribute, with 'Is_Image coming in second.

This could be interesting for enum values that have constraints - either a range constraint or a predicate that mark certain values as invalid, because the image is Valid for the enum's 'Base type, in that case.  But 'Valid would still return False for values omitted by a range constraint or a predicate, so we'd want this proposed attribute to return False for this cases too, or it couldn't serve the intent.

I agree this would be useful.
Without it, we can just catch the exception, which is usually an exceptional case, but there are occasions when it would be nice to check before calling 'Value.

**Comment by ArnaudCharlet** [comment]
> between overloading 'Valid or using 'Valid_Image, 'Is_Image, or 'Fred,
> I prefer 'Valid_Image for the name of the proposed attribute, with 'Is_Image coming in second.
> 
> This could be interesting for enum values that have constraints - either a range constraint or a predicate that mark certain values as invalid, because the image is Valid for the enum's 'Base type, in that case.  But 'Valid would still return False for values omitted by a range constraint or a predicate, so we'd want this proposed attribute to return False for this cases too, or it couldn't serve the intent.
> 
> I agree this would be useful.
> Without it, we can just catch the exception, which is usually an exceptional case, but there are occasions when it would be nice to check before calling 'Value.

FWIW, GNAT has a 'Valid_Value attribute for that purpose: https://docs.adacore.com/gnat_rm-docs/html/gnat_rm/gnat_rm/implementation_defined_attributes.html#attribute-valid-value

**Comment by CKWG** [comment]
> I think "Valid_Image", or "Is_Image", or even "Fred" would be a better name for this attribute.

I prefer 'Valid_Value to 'Fred or even 'Trump.

**Comment by AJ-Ianozi** [comment]
I also agree that this would be incredibly useful to have and if GNAT is already using `'Valid_Value` for this purpose I'd just opt for using that.

**Comment by joshua-c-fletcher** [comment]
I think `'Valid_Image` makes sense because it is checking if the string is a valid image of an enum value, and `'Valid_Value` makes sense because it is checking if the string can be converted via `'Value` to a valid value.

The latter is actually is actually easier to evaluate: since `'Image` can be customized via `'Put_Image` the concepts aren't as similar as they used to be.  If not for that, I'd have thought `'Valid_Image` the more intuitive, but custom images make that harder to guarantee, so `'Valid_Value` is a better name for that reason, now.

Given that GNAT already has `'Valid_Value`, I agree it makes sense to use that for that reason as well.  It would be strange to strange to have GNAT needing to keep `'Valid_Value` for backwards compatibility while providing almost the same thing with `'Valid_Image`, for example.

I do find something strange about GNAT's definition, though.  According to GNAT's definition, it currently only applies to enumerations that are not defined in the package Standard: https://docs.adacore.com/gnat_rm-docs/html/gnat_rm/gnat_rm/implementation_defined_attributes.html#attribute-valid-value

It seems odd and arbitrary to omit the enumerations defined in the package Standard, and that limits the usability of this attribute from within a generic.  GNAT allows you to write a generic that uses this attribute, but it won't let you instantiate it if the discrete type using this attribute in the generic happens to be a Boolean or a Character, for example, or a non-enum discrete type (since the most specific you can get with generic formal parameters is that the type is discrete - could be an enum, range or mod type).

I don't think there's any precedence in the Ada language for discrete type attributes that apply _only_ to enumerations that are _not_ in the package Standard.  If we make `'Valid_Value` part of the language (and I think that _is_ a good idea), I think the restriction should be relaxed so that it can be used with enums in standard, and perhaps more generally for Scalar types, as that is the scope of the `'Value` attribute it relates to.

**Comment by CKWG** [comment]
Hm, the only enums in Standard are Boolean and [[Wide_]Wide_]Character. I doubt that the attribute will ever be needed for Boolelan. On the other hand, the nonidentifiers in Character make it difficult to define it.
I do see the problem with generics.

For scalar types, there is this exception in RM:
A.10.10(17/1) Although the specification of the generic package Enumeration_IO would allow instantiation for an integer type, this is not the intended purpose of this generic package, and the effect of such instantiations is not defined by the language.

**Comment by joshua-c-fletcher** [comment]
With respect to what you called the 'nonidentifiers in Character', any new enum can include character-like values, along with other more typical enum values, so [[Wide_]Wide_]Character is not unique in that respect.
`type Example is ('a', to, 'z');`
As for Boolean, the attribute could be used in a generic, and I can imagine a cases where such a generic might be used with Booleans and other enums.

One potential issue with having it only defined for enumerations and not scalar types is that its use in a generic immediately makes it impossible to instantiate the generic if the type is not an enumeration, and it is impossible to specify that the generic should only take an enumeration for that parameter (except via comments and naming conventions).

Generally, it's not necessary to distinguish enums from other discrete types in a generic formal parameter, so it hasn't been a problem.  For Enumeration_IO, the clarity is in the name of the package, but if enums become more distinct from other discrete types, there might be a need to specify that a generic should only take an enumeration (rather than any discrete type)... and I don't think we need or want to add a way to specify that when they're otherwise similar enough to other discrete types.  

There's also no way within a generic to check if a discrete type is an enum or other discrete type (except perhaps to check the image of an arbitrary value - by default other discrete type images would start with a space or negative sign, and an enum image never would... but that's a clunky check, generally not necessary, and less guaranteed with `'Put_Image` overrides), so I don't generally think we should need to check if a discrete type is an enum... but if there are attributes unique to enums that are only available in a generic if it happens to be instantiated with an enum, it makes it a bit more complicated (and I think unnecessarily so).  Removing that restriction would eliminate the problem and the complexity.

**Comment by CKWG** [comment]
> With respect to what you called the 'nonidentifiers in Character', any new enum can include character-like values, along with other more typical enum values, so [[Wide_]Wide_]Character is not unique in that respect.
> `type Example is ('a', to, 'z');`

You misunderstood. `Example'(to)` is an identifier. What I meant is the nonidentifiers in type Character denoted in italics:
_nul, soh, stx, ..._, which are nonidentifiers, you cannot name them.

For the rest, I said
> I do see the problem with generics.

**Comment by ARG-Editor** [comment]
> One potential issue with having it only defined for enumerations
> and not scalar types is that its use in a generic immediately makes
> it impossible to instantiate the generic if the type is not an
> enumeration, and it is impossible to specify that the generic
> should only take an enumeration for that parameter (except via
> comments and naming conventions).

That's not how generics in Ada work (can't speak for GNAT). One cannot
reject an instantiation because of anything that occurs in the body; either
the body itself has to be illegal or the instantiation has to be accepted.
The usual work-around is to raise Program_Error if something in the body
doesn't make sense given the actual. But we really try to avoid such things.

Given that 'Value is defined on all scalar types, it would make the most
sense for 'Valid_Value to be defined on the same types. It seems bizarre to
only allow it on some sort of subset of the types that Value can be used
with. It seems equally useful to find out if a numeric literal has the
correct form.

       Randy.

**Comment by joshua-c-fletcher** [comment]
Sounds like we agree.

**Comment by ARG-Editor** [comment]
The Valid_Value attribute is now written up in AI22-0172-1; it will be on the agenda for a future ARG meeting.
